File: KoKoslan/src/java/ast/KoKoBiOperation.java
/**
 @author loriacarlos@gmail.com
 
*/
package kokoslan.ast;
import java.util.*;
import java.io.*;


public class KoKoBiOperation extends KoKoOperation {
	
    public KoKoBiOperation(KoKoAst operator, KoKoAst left, KoKoAst right){
		super(operator, Arrays.asList(left, right));
	}
	
	public KoKoAst left(){return this.operands.get(0);}
	public KoKoAst right(){return this.operands.get(1);}
	
	@Override
	public KoKoValue eval(KoKoContext ctx){
	   try {
	        KoKoId operId = (KoKoId)operator;
			KoKoNumValue lv = (KoKoNumValue)(left().eval(ctx));
			KoKoNumValue rv = (KoKoNumValue)(right().eval(ctx));
			switch (operId.getValue()){
				case "+" : return new KoKoNumValue(lv.getValue() + rv.getValue());
				case "-" : return new KoKoNumValue(lv.getValue() - rv.getValue());
				case "*" : return new KoKoNumValue(lv.getValue() * rv.getValue());
				default  : throw new KoKoEvalException("KoKoBiOperation: unimpemented operator");
			}
			
	   } catch (Exception e) {
			throw new KoKoEvalException(e.getMessage());
	     }
	   
   }
	
}

File: KoKoslan/src/kotlin/ast/KoKoBiOperation.kt
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.kt.ast
import java.util.*
import java.io.*
import kokoslan.kt.eval.*
import kokoslan.kt.exception.*


open class KoKoBiOperation(oper: KoKoAst, left: KoKoAst, right: KoKoAst): KoKoOperation(oper, Arrays.asList(left, right)) {

    fun left(): KoKoAst {
        return this.operands.get(0)
    }
    fun right(): KoKoAst {
        return this.operands.get(1)
    }

    override fun eval(ctx: KoKoContext): KoKoValue {
        try {
            val operId = this.oper as KoKoId
            val lv = left().eval(ctx) as KoKoNumValue
            val rv = right().eval(ctx) as KoKoNumValue
            when(operId.getValue()) {
                "+" -> return KoKoNumValue(lv.getValue() + rv.getValue())
                "-" -> return KoKoNumValue(lv.getValue() - rv.getValue())
                "*" -> return KoKoNumValue(lv.getValue() * rv.getValue())
                else -> throw KoKoEvalException("KoKoBiOperation unimpemented operator")
            }

        } catch (e: Exception) {
            throw KoKoEvalException(e.message!!)
        }

    }

}